README.md

```diff
@@ -130,75 +129,74 @@ const login = function (request, reply) {
             (message ? '<h3>' + message + '</h3><br/>' : '') +
             '<form method="post" action="/login">' +
             'Username: <input type="text" name="username"><br>' +
             'Password: <input type="password" name="password"><br/>' +
-            '<input type="submit" value="Login"></form></body></html>');
+            '<input type="submit" value="Login"></form></body></html>';
     }
 
     const sid = String(++uuid);
-    request.server.app.cache.set(sid, { account: account }, 0, (err) => {
 
-        if (err) {
-            reply(err);
-        }
+    await request.server.app.cache.set(sid, { account }, 0);
+    request.cookieAuth.set({ sid });
 
-        request.cookieAuth.set({ sid: sid });
-        return reply.redirect('/');
-    });
+    return h.redirect('/');
 };
 
-const logout = function (request, reply) {
+const logout = function (request, h) {
 
     request.cookieAuth.clear();
-    return reply.redirect('/');
+    return h.redirect('/');
 };
 
-const server = new Hapi.Server();
-server.connection({ port: 8000 });
+const server = new Hapi.Server({ port: 8000 });
 
-server.register(require('../'), (err) => {
+exports.start = async () => {
 
-    if (err) {
-        throw err;
-    }
+    await server.register(require('../'));
 
     const cache = server.cache({ segment: 'sessions', expiresIn: 3 * 24 * 60 * 60 * 1000 });
     server.app.cache = cache;
 
-    server.auth.strategy('session', 'cookie', true, {
+    server.auth.strategy('session', 'cookie', {
         password: 'password-should-be-32-characters',
         cookie: 'sid-example',
         redirectTo: '/login',
         isSecure: false,
-        validateFunc: function (request, session, callback) {
-
-            cache.get(session.sid, (err, cached) => {
+        validateFunc: async function (request, session) {
 
-                if (err) {
-                    return callback(err, false);
-                }
+            const cached = await cache.get(session.sid);
+            const out = {
+                valid: !!cached
+            };
 
-                if (!cached) {
-                    return callback(null, false);
-                }
+            if (out.valid) {
+                out.credentials = cached.account;
+            }
 
-                return callback(null, true, cached.account);
-            });
+            return out;
         }
     });
 
+    server.auth.default('session');
+
     server.route([
         { method: 'GET', path: '/', config: { handler: home } },
         { method: ['GET', 'POST'], path: '/login', config: { handler: login, auth: { mode: 'try' }, plugins: { 'hapi-auth-cookie': { redirectTo: false } } } },
         { method: 'GET', path: '/logout', config: { handler: logout } }
     ]);
 
-    server.start(() => {
+    await server.start();
 
-        console.log('Server ready');
-    });
-});
+    console.log('Server ready');
+};
+
+try {
+    exports.start();
+}
+catch (err) {
+    console.log(err.stack);
```

Comment:
i feel like there should be a `process.exit(1);` here, as-is this would log an error but exit with a non-error exit code, that seems bad

Reply:
I doubt the catch would ever be triggered since node doesn't have top level async/await.

Reply:
good point, it would be a promise rejection, so `exports.start().catch((err) => {});` is more appropriate

Reply:
So what do you all think about wrapping this in a self-executing async function?

Reply:
Also, I didn't really want to make too large of a change in this PR aside from making it compatible with hapi17, but I'm not a huge fan of the fact the whole `example/index.js` is in the readme. I feel like I could just keep the example updated and point to it from here.